<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/css/bootstrap.min.css" integrity="sha384-/Y6pD6FV/Vv2HJnA6t+vslU6fwYXjCFtcEpHbNJ0lyAFsXTsjBbfaDjzALeQsN6M" crossorigin="anonymous">
        <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css"
              integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU"
              crossorigin="anonymous">
        <link rel="stylesheet" type="text/css" href="../css/dendrogram.css">
        <link rel="stylesheet" type="text/css" href="../css/expressMap.css">
        <link rel="stylesheet" type="text/css" href="../css/isoform.css">
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Libre+Franklin:300,400" >
        <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.4/css/select2.min.css">

    </head>

    <body class="container-fluid" style="margin-top:20px; margin-left: 20px;">
    <h3>GTEx Transcript Browser</h3>
    <hr>
        <div class="row">
            <div class="col-xs-12 col-sm-2">
                <form id="isoformExpressForm" style="padding: 5px;">
                     <div class="form-group"> <!-- form-group for optimum spacing between the label and its associated form input element -->
                        <input id="gene" type="text" class="form-control" placeholder="Enter a gencode ID or gene name" style="font-size:12px; text-transform:uppercase;"></input>
                         <small>e.g. ACTN1, ACTN3, NDRG4, ENGASE, PTENP1 (a single-exon gene), ACTN1-AS1 (a single-exon gene)</small>
                     </div>
                </form>
            </div>
            <div id="InfoBox" class="col-xs-12 col-sm-10">
                <div><i id="spinner" class="fas fa-sync fa-spin" style="display:none;"></i></div>
                <div style="font-size:12px;">
                    <p>
                        GTEx Transcript Browser is for exploring expression of individual transcripts of a gene.
                        The top tab bar in the browser provides view options for browsing expression data at
                        the exon, junction or transcript level across multiple tissues summarized in a heatmap.
                    </p>

                    <p>
                        In the transcript expression view, the expressions of transcripts across tissues are summarized
                        in a heatmap. The tissues (rows) and isoforms (columns) are hierarchically clustered using
                        the Euclidean distance and average linkage method. The dendrogram scale shows the cluster
                        distance. The individual transcript structures are illustrated below the heatmap--exons are
                        represented as rectangles, and introns (the distance between two adjacent exons) are
                        scaled by length up to 1 kilobases (kb) long. By clicking a tissue label on the heatmap,
                        a lollipop plot of the transcript expression in the selected tissue will appear next to
                        the transcript tracks.
                    </p>

                    <p>
                        Exon-level expression is calculated based on a collapsed gene model. To learn more about
                        the collapsing transcript model, go to GTEx Portal Documentation: Analysis Methods.
                        In the exon view, the heatmap summarizes expression of exons of the collapsed gene model across
                        tissues. The tissues (rows) are hierarchically clustered based on the expression data.
                        The exons (columns) are ordered by genomic location. Below the heatmap, the collapsed gene model
                        is illustrated--exons are represented as rectangles, and introns are scaled by length up to 1 kb long.
                        Additionally, the junctions between joined exons of all splice variants are mapped out on the
                        model and shown as arcs. By clicking a tissue label on the heatmap, the gene model will change
                        colors scaled by expression in the selected tissue. At times, an exon on the gene model may
                        show dashed-line borders indicating that the exon or part of it has been filtered in the collapsed model.
                    </p>

                    <p>
                        In the junction view, the heatmap summarizes expression of junctions from all transcripts across
                        tissues. The tissues (rows) are hierarchically clustered based on the junction expression data,
                        and the junctions (columns) are sorted by genomic location.</p>

                    <p>
                        All expression quantifications are described in the GTEx Portal Documentation: Analysis Methods
                        -> Expression Quantification section of the portal. </p>

                </div>

            </div>
        </div>
        <div id="searchResult" class="col-xs-12">
                <div id="messageBox" style="color: #CB181D;"></div>
                <!-- charts -->
                <div id="isoformNav" style="display:none">
                    <!-- Nav tabs -->

                    <ul class="nav nav-tabs" role="tablist">
                        <li id="defaultTab" class="nav-item">
                            <a id="exon" class="nav-link active" role="tab" data-toggle="tab">Exon Expression</a>
                        </li>

                        <li class="nav-item">
                            <a id="junction" class="nav-link" role="tab" data-toggle="tab">Junction Expression</a>
                        </li>

                        <li class="nav-item">
                            <a id="isoform2" class="nav-link" role="tab" data-toggle="tab">Transcript Expression</a>
                        </li>

                    </ul>

                    <!-- Tab panes -->
                    <div class="tab-content">
                        <div id="vizRoot" style="width:80%;"></div> <!-- the heatmap/gene model Div -->
                    </div>
                </div>

        </div>

    </body>
    <script
      src="https://code.jquery.com/jquery-3.2.1.min.js"
      integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4="
      crossorigin="anonymous">
    </script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.11.0/umd/popper.min.js" integrity="sha384-b/U6ypiBEHpOf/4+1nzFpr53nxSS+GLCkfwBdFNTxtclqqenISfwAzpKaMNFNmj4" crossorigin="anonymous"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/js/bootstrap.min.js" integrity="sha384-h0AbiXch4ZDo7tp9hKZ4TsHbi047NrKGLO3SEJAg45jXxnGIfYzk4Si90RDIqNm1" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.4/js/select2.full.min.js"></script>

    <script src="../build/js/transcript-browser.bundle.dev.js"></script>
    <script>
        $(document).ready(function(){
            $('#gene').keypress(function(e){
                if(e.keyCode === 13){
                    // bind the enter key
                    e.preventDefault(); // Note: prevent the default behavior of the enter key, which is refreshing the page
                    init("exon");
                    $('.nav-link').removeClass("active");
                    $('#exon').addClass("active");

                }
            });
            $('#isoform').click(function(){init("isoform")});
            $('#isoform2').click(function(){init("isoformTransposed")});
            $('#exon').click(function(){init("exon")});
            $('#junction').click(function(){init("junction")});

        });
        function init(type){
            const geneId = $('#gene').val();
            reset();
            $('#spinner').show();
            $('#isoformNav').show();
            TranscriptBrowser.render(type, geneId, "vizRoot");
        }
        function reset(){
            $("#vizRoot").empty();
        }
    </script>

</html>
